src/ir/auth_logic/ast_construction_test.cc

```diff
@@ -26,51 +26,76 @@
 #include <variant>
 #include <vector>
 
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_join.h"
 #include "src/common/testing/gtest.h"
 #include "src/ir/auth_logic/ast.h"
 #include "src/ir/datalog/program.h"
 
 namespace raksha::ir::auth_logic {
 
+using ::testing::UnorderedElementsAre;
+
 struct QueryTestData {
   std::string input_text;
   std::string query_name;
-  std::string principal_name;
-  std::string fact_name;
-  datalog::Predicate predicate;
+  Principal principal;
+  Fact fact;
 };
 
+std::string ToString(datalog::Predicate predicate) {
+  return absl::StrCat(predicate.name(), "(",
+                      absl::StrJoin(predicate.args(), ", "), ")");
+}
+std::string ToString(BaseFact basefact) {
+  if (basefact.GetValue().index() == 0) {  // Predicate BaseFact
```

Comment:
Don't use `index` on a `std::variant`. Imagine someone removes or adds a variant from the front of the list: all of your indices will get messed up and mean something different, and you'll have to fix every index in the program.

Instead, you should just use `visit` with a lambda to get these alternatives. If every `Datalog` AST structure must have a `ToPrettyString` method, you can just do this:

```C++
// We are within the method `ToPrettyString` in `BaseFact`, so we have access to value_.
std::visit(value_, [](const auto &arg) { arg.ToPrettyString(); }
```

Reply:
Removed the usage of index and added std::visit(